snuba/query/processors/null_column_caster.py

```diff
@@ -0,0 +1,125 @@
+    >>> LIMIT 1
+    >>> OFFSET 0
+
+    And clickhouse will not throw an error since the column will be interpreted as nullable
+
+
+    """
+
+    def _find_mismatched_null_columns(self) -> Dict[str, FlattenedColumn]:
+        mismatched_col_name_to_col: Dict[str, FlattenedColumn] = {}
+        col_name_to_nullable: Dict[str, bool] = {}
+        for table_storage in self.__merge_table_sources:
+            for col in table_storage.get_schema().get_columns():
+                col_is_nullable = _col_is_nullable(col)
+                other_storage_column_is_nullable = col_name_to_nullable.get(
+                    col.name, None
+                )
+                if (
+                    other_storage_column_is_nullable is not None
+                    and other_storage_column_is_nullable != col_is_nullable
+                ):
+                    mismatched_col_name_to_col[col.name] = col
+                col_name_to_nullable[col.name] = col_is_nullable
+
+        return mismatched_col_name_to_col
+
+    def __init__(self, merge_table_sources: Sequence[ReadableTableStorage]):
+        """
+        Args:
+            merge_table_sources: sequence of the storages which make up the merge table,
+            This is necessary to find which fields need to be cast to nullable
+
+        """
+        self.__merge_table_sources = merge_table_sources
+        self.__mismatched_null_columns = self._find_mismatched_null_columns()
+
+    @property
+    def mismatched_null_columns(self) -> Dict[str, FlattenedColumn]:
+        return self.__mismatched_null_columns
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        def cast_column_to_nullable(exp: Expression) -> Expression:
+            if isinstance(exp, Column):
+                if exp.column_name in self.mismatched_null_columns:
+                    # depending on the order of the storage, this dictionary will contain
+                    # either the nullable or non-nullable version of the column. No matter
+                    # which one is in there, due to the mismatch on the merge table it needs to
+                    # be cast as nullable anyways
+                    mismatched_column = self.mismatched_null_columns[exp.column_name]
+                    col_is_nullable = _col_is_nullable(mismatched_column)
+                    col_type = mismatched_column.type.for_schema()
+                    cast_str = col_type if col_is_nullable else f"Nullable({col_type})"
+                    return FunctionCall(
+                        exp.alias,
+                        "cast",
+                        (
+                            # move the alias up to the cast function
+                            Column(
+                                None,
+                                table_name=exp.table_name,
+                                column_name=exp.column_name,
+                            ),
```

Comment:
I am not sure about this move here and want to make sure the reviewer sees this line. I do not know how our system uses aliases and what I should do with the existing alias (if anything)

Reply:
What you're doing is correct. We want to ensure that the alias of the expression going in to your function is the alias of the expression coming out. Those aliases might be used elsewhere in the query. You don't really need to alias the inner `Column` so that's also fine.